api/capi/src/nntrainer.cpp

```diff
@@ -231,45 +253,62 @@ int ml_train_model_destroy(ml_train_model_h model) {
     NN->finalize();
     return ML_ERROR_NONE;
   };
-
   status = nntrainer_exception_boundary(f);
 
-  if (nnmodel->optimizer)
+  if (nnmodel->optimizer) {
+    ML_TRAIN_RESET_VALIDATED_HANDLE(nnmodel->optimizer);
     delete nnmodel->optimizer;
-  if (nnmodel->dataset)
+  }
+
+  if (nnmodel->dataset) {
+    ML_TRAIN_RESET_VALIDATED_HANDLE(nnmodel->dataset);
     delete nnmodel->dataset;
-  for (auto &x : nnmodel->layers_map)
+  }
+
+  for (auto &x : nnmodel->layers_map) {
+    ML_TRAIN_RESET_VALIDATED_HANDLE(x.second);
     delete (x.second);
+  }
   nnmodel->layers_map.clear();
+
   delete nnmodel;
 
   return status;
 }
 
-int ml_train_model_get_summary(ml_train_model_h model,
-                               ml_train_summary_type_e verbosity,
-                               char **summary) {
-  if (summary == nullptr) {
-    ml_loge("summary pointer is null");
-    return ML_ERROR_INVALID_PARAMETER;
-  }
-
+static int ml_train_model_get_summary_util(ml_train_model_h model,
+                                           ml_train_summary_type_e verbosity,
+                                           std::stringstream &ss) {
   int status = ML_ERROR_NONE;
   ml_train_model *nnmodel;
 
-  ML_NNTRAINER_CHECK_MODEL_VALIDATION(nnmodel, model);
+  ML_TRAIN_GET_VALID_MODEL_LOCKED(nnmodel, model);
```

Comment:
I don't think we need to lock when we just display summary of network.

Reply:
We need lock for summary as well because a user can delete the network while getting the summary.